### ISSUE TITLE
CRF implementation in Keras is not  not giving  good results

### ISSUE THREAD

**Issue by OpenCv30** [baidu-research/NCRF#34]
Hi,
I am a student and working on Camelyon’16 as my Master’s project. 
I was going through your very impressive paper - **Yi Li and Wei Ping. Cancer Metastasis Detection With Neural Conditional Random Field. Medical Imaging with Deep Learning (MIDL), 2018.**
And, found that you have implemented CRF in your code on top of Resnet 18. So far I am using Resnet50 but my FROC score is not going up from 0.55.

So, I have decided to use your approach and re-implemented your code in Keras (backend-Tensorflow). But the performance of the trained model is not even close to your results. Best **FROC of Resnet18+CRF trained model-0.55**. Lot of FPs are coming. My Resnet18 is taken from https://github.com/raghakot/keras-resnet

My queries-

- **My training loss- BCE- started from 1.16 and finally settled to 0.8639 and validation loss 0.8528**. It is right or loss should go further down. I have run for  30+ epochs, but the loss remains the same(plateau). Don’t know why? (please refer attached image below)

- Weight plot doesn’t look closer to what you have shown in your paper. In your, case all the positional patch weights W[0,0,0,0] <0,W[0,1,0,1] <0..W[2,2,2,2] <0 but I am not seeing that. As per equation, these will not affect the final predictions. (please refer attached image below)

Can you help me in understanding why loss is not going further down? It gets plateau after a certain number of epochs and after that no effect even with Cyclic Learning rate [ 1e-4,1e-5,1e-7]. This behavior is common across many models – rennet50/101/18 + Inception V3.
Please help me to solve these problems. I shall be thankful to you.

**Training Configration** 
![config](https://user-images.githubusercontent.com/13656983/56684919-5670e680-66ee-11e9-954e-396742676d5a.PNG)

**TensorBoard ACC/LOSS plot**
![image002](https://user-images.githubusercontent.com/13656983/56684692-ebbfab00-66ed-11e9-9d68-5c4e4c59dc68.jpg)

![image006](https://user-images.githubusercontent.com/13656983/56684734-fe39e480-66ed-11e9-91cb-7d57b1c25f25.jpg)

**Just Validation loss and training loss ( after 16 epochs )  -- Orange is val-loss, Blue is Training  BCE loss**
![val_train_loss](https://user-images.githubusercontent.com/13656983/56685948-af417e80-66f0-11e9-84c7-6ddf8ee0f79a.PNG)

**Weight plots-**
**plot across epochs-**

![weight_epochs](https://user-images.githubusercontent.com/13656983/56687604-570c7b80-66f4-11e9-80f2-a2b560e0c863.PNG)

**Just one of the epoch(16)  weight map from which heatmap is generated**
![image009](https://user-images.githubusercontent.com/13656983/56684808-21fd2a80-66ee-11e9-9077-85c1f29a5e96.jpg)

**Heatmaps**
Test_001.tiff ( cam'16 test data set)
**Results from my trained model- at level 8**

![001](https://user-images.githubusercontent.com/13656983/56685356-65a46400-66ef-11e9-836a-b7e9b4be3d88.PNG)

**Results for your model-at level 6**
![test_001_crfbaidu](https://user-images.githubusercontent.com/13656983/56685139-dbf49680-66ee-11e9-8c58-344ce3e5845b.png)

Clearly, your model performs far better than my trained model.

I have matched my CRF implementation in kears with yours in pyTorch. For the same input in both the model, I am getting the same output.

Please help me to reproduce your results in Kears+ TF.

**Comment by yil8** [comment]
@OpenCv30 Hi, so sorry for the late reply. First of all, did you only try resenet50 through my repo? There are other users using the default config resnet18 and were able to reproduce my results. I would definitely recommending reproducing my results through my repo first before trying re-implementing yourself.

**Comment by OpenCv30** [comment]
Thank for your reply. 
I am using resnet50 architecture from keras without imagenet weight as i am not good at pytorch.
I have reproduced the heatmap of Test026.tif which is matching to your’s shared one. So, there is no problem with the model checkpoints given by you. 
By the ways, it this normal behaviour of training that initially loss falls down and later remains same with any change? 
Please suggest what results you like to see or i should try to improve performance of my model. If you want then i can share my implementation of CRF layers code in Kears.

**Comment by yil8** [comment]
@OpenCv30 Typically, deep neural networks will overfit training data essentially, so I would expect training loss approaching 0 if it's not a gigantic dataset. Sorry, but I guess I don't have enough time to look into your code/implementation of CRF and check if it's correctly implemented. On the other hand, I thought pytorch is pretty easy to pick up...

**Comment by OpenCv30** [comment]
@yil8  As you have suggested, first I will learn pytorch to train a model from your code directly and generated heatmaps for Cam'16 test data to get desired Froc. But as such, I have never seen a BCE loss coming close to zeros for Camelyon Dataset.  
I understand that you might not have time to see my code. But, can you please tell me from my training loss/acc profiles and weight maps that training is correct or not. My understanding is that initially model was getting trained as CRF weights are changing but later CRF weights got stagnant. Might be vanishing gradient issue. Does it make sense to use this model to generate Hard negatives and train again even though your 200000 points has hard negatives?

I will get back to you soon please don't close this issue.

**Comment by yil8** [comment]
@OpenCv30 That's indeed weird that you can't make your BCE loss goes to zero on the training data. Can you turn off CRF and just do purely CNN and see if you can make your BCE loss goes to zero? Then you know if you may have bugs either in the CNN part or the CRF part.

**Comment by OpenCv30** [comment]
@yil8 - As you have suggested, I started training only resnet18 with your the patch locations shared by you.
I am seeing again loss is plateaued near 0.88. It is not going down even after 40 epochs. 
![onlyResnet18](https://user-images.githubusercontent.com/13656983/57209503-575f1d80-6ff6-11e9-9a06-a5a6990104b9.png)

Please suggest me what could be wrong.  I have used the same training framework where I was seeing loss reaching around 0.10-0.15 before the plateau.

**Comment by yil8** [comment]
@OpenCv30 are you training with my code or with your code? It's also strange that you have a significantly lower val-loss than train-loss, which almost never happened in my life...
